Add shortened workspace path when displaying recent workspaces

vince-fugnitto opened the pull request:
Updated the workspace path when displaying recent workspaces on non-windows operating systems.
Instead of displaying the full path `ex: /home/[username]/` replace it with shortened notation `~/`.

For workspaces which are on the home folder, display the full path.

Signed-off-by: Vincent Fugnitto <vincent.fugnitto@ericsson.com>

vince-fugnitto commented:
one issue I might have is whether or not the `workspace-service` is the correct provider to obtain information on `platform` and `homedir` or if there is a more suitable service.

simark commented:
It works great for me.  I don't know where to put `platform` and `homedir`, though.

simark commented on packages/workspace/src/browser/quick-open-workspace.ts:
```diff
@@ -75,12 +80,15 @@ export class WorkspaceQuickOpenItem extends QuickOpenItem {
     getDescription(): string {
-        return (this.workspace) ? new URI(this.workspace).path.dir.toString() : '';
+        let path = (this.workspace) ? new URI(this.workspace).path.dir.toString() : '';
+        if (this.platform !== 'win32' && path.includes(this.home)) {
```
`includes`?  What if my home is `/home/simark` and the workspace is `/backups/pc/home/simark/foo`?

simark replied:
Another corner-case test to add: path `/home/simark123/foo` and home `/home/simark`.

vince-fugnitto replied:
nice catch! I should only replace the result of the user's `home` if it is at the start of the path.

simark commented on packages/workspace/src/browser/workspace-service.ts:
```diff
@@ -106,6 +106,14 @@ export class WorkspaceService implements FrontendApplicationContribution {
         return !!this._root;
     }
 
+    get home(): Promise<string> {
```
Please document these new methods.

akosyakov commented on packages/workspace/src/browser/quick-open-workspace.ts:
```diff
@@ -30,10 +30,13 @@ export class QuickOpenWorkspace implements QuickOpenModel {
-    open(workspaces: string[]): void {
+    async open(workspaces: string[]): Promise<void> {
         this.items = [];
+        const home = await this.workspaceService.home;
```
https://github.com/theia-ide/theia/blob/a821f30a4961913bd09fc5793bf2cf9114d525a6/packages/filesystem/src/common/filesystem.ts#L115

akosyakov commented on packages/workspace/src/browser/quick-open-workspace.ts:
```diff
@@ -30,10 +30,13 @@ export class QuickOpenWorkspace implements QuickOpenModel {
+    async open(workspaces: string[]): Promise<void> {
         this.items = [];
+        const home = await this.workspaceService.home;
+        const platform = await this.workspaceService.platform;
```
https://github.com/theia-ide/theia/blob/a821f30a4961913bd09fc5793bf2cf9114d525a6/packages/core/src/common/os.ts#L29

svenefftinge replied:
It needs to get the backend's platform. But still should not be added to `workspaceService`. Can't we detect that from the URI somehow?

vince-fugnitto replied:
@svenefftinge @akosyakov perhaps I can create a service in core named `HostService` which provides methods related to information from the host server, and can be further updated to include any necessary helper methods.

Example: `getPlatform(), isHostWindows(), isHostMac(), getHomedir(), ...`

vince-fugnitto commented:
@akosyakov I made changes based on the helpful information you provided regarding `home`, and `platform`

simark commented on packages/workspace/src/browser/quick-open-workspace.ts:
```diff
@@ -75,12 +83,18 @@ export class WorkspaceQuickOpenItem extends QuickOpenItem {
+        const path = uri.path.dir.toString();
+        const isWindows = (uri.path.root) ? Path.isDrive(uri.path.root.base) : undefined;
+
+        if (!isWindows && this.home && this.homeRegex.test(path)) {
```
Is there a reason for using a regex, and not just looking if the string starts with `${this.home}/`?  This should work:

```ts
        if (!isWindows && this.home && path.indexOf(`${this.home}/`) === 0) {
            return path.replace(`${this.home}/`, '~/');
        }
```

It's not very likely, but if the home path contains characters that have a special meaning in regexes (e.g. `/home/[simark]`), it probably won't work correctly.  And since we don't need the extra power of regexes here, we might as well use the simple solution.

simark commented on packages/workspace/src/browser/quick-open-workspace.ts:
```diff
@@ -28,11 +29,15 @@ export class QuickOpenWorkspace implements QuickOpenModel {
-    open(workspaces: string[]): void {
+    async open(workspaces: string[]): Promise<void> {
         this.items = [];
+        const home = await this.fileSystem.getCurrentUserHome().then(stat => stat ? new URI(stat.uri).withoutScheme().toString() : undefined);
```
If you use the `.then(...)` form, you also need a `.catch(...)`.  You have probably seen this message before:

```
(node:12964) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 2): 1
(node:12964) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```
This means that if `this.fileSystem.getCurrentUserHome()` fails, we'll get this warning, and in future versions of node it will kill the process.  I would suggest doing it in two steps like this:

```
        const homeStat = await this.fileSystem.getCurrentUserHome();
        const home = homeStat ? new URI(homeStat.uri).withoutScheme().toString() : undefined;
 ```

simark commented:
Could you try to do a test for `WorkspaceQuickOpenItem.getDescription`? There are a few corner cases that would be worth testing, such as (with `/home/foo` as the home):

* `/home/foo2/workspace`
* `/home/foo/hello/home/foo/workspace`
* `/tmp/home/foo/workspace`

and maybe more.

simark commented on packages/core/src/common/workspace-utils.ts:
```diff
@@ -0,0 +1,37 @@
+     * @param workspace
+     * @param home
+     */
+    export function getShortenedPath(workspace: string, home?: string): string {
```
Rename the workspace variable to something else (like `path`), since it doesn't have to be a workspace path, it could the path of anything.

simark replied:
I would suggest making the `home` parameter not `?`, since calling this function with `home == undefined` is pointless.  Though we still need to handle the case where the caller passes an empty string as the home.

simark commented on packages/core/src/common/workspace-utils.ts:
```diff
@@ -0,0 +1,37 @@
+export namespace WorkspaceUtils {
+
+    /**
+     * Returns shortened workspace path if applicable
```
It's a bit vague what "shortened workspace path" means, the function name `getShortenedPath` is also not very clear.  What about `makeTildePath` or something like that?  The doc could be:

    If `path` is a descendant of `home`, replace the home directory part of `path` with `~`.  This function is a no-op on Windows.

simark commented on packages/core/src/common/workspace-utils.ts:
```diff
@@ -0,0 +1,37 @@
+     */
+    export function getShortenedPath(workspace: string, home?: string): string {
+        const path = new Path(workspace);
+        const dirPath = path.dir.toString();
```
I think this function should not strip away the last portion of the path.  If I call

    getShortenedPath("/home/simark/something", "/home/simark")

I should get back `"~/something"`.

simark commented on packages/core/src/common/workspace-utils.spec.ts:
```diff
@@ -0,0 +1,80 @@
+import { expect } from 'chai';
+import { WorkspaceUtils } from './workspace-utils';
+
+describe('workspace-service', () => {
```
It would be good to test the case when home is an empty string.  Your implementation looks fine, but a change in the future could easily lead to:

    getShortenedPath("/home/simark/foo", "") -> "~/home/simark/foo"

which would be wrong.

I'm thinking about other cases, though I don't think we really need to handle them:

    getShortenedPath("/home/simark/foo", "/home/simark/")
    getShortenedPath("/home/vince/../simark/foo", "/home/simark")

vince-fugnitto commented:
@simark I made changes based on the helpful feedback you provided. 
- refactored `tildifyPath` function name along with it's parameter names.
- refactored `home` param to become mandatory.
- updated `tildifyPath` to not strip trailing folder name. 
(ex: `/home/vincent/Workspaces/theia` to `~/Workspaces/theia`)
- added an additional test case to handle when `home` is empty.

simark commented on packages/core/src/common/workspace-utils.spec.ts:
```diff
@@ -0,0 +1,92 @@
+    const linuxHome = '/home/test-user';
+    const macHome = '/Users/test-user';
+    const windowsHome = '/C:/Users/test-user';
+    const name = 'theia';
```
I guess it's not really useful to have that `name` variable.

simark commented on packages/core/src/common/workspace-utils.spec.ts:
```diff
@@ -0,0 +1,92 @@
+
+describe('workspace-service', () => {
+    const linuxHome = '/home/test-user';
+    const macHome = '/Users/test-user';
```
Is there a reason to have "linux" and "mac" tests?  They are not really different, the home is just an arbitrary path.

simark commented on packages/core/src/common/workspace-utils.ts:
```diff
@@ -0,0 +1,39 @@
+     * @param resourcePath
+     * @param home
+     */
+    export function tildifyPath(resourcePath: string, home: string): string {
```
Haha, I like the name :)

simark commented on packages/workspace/src/browser/quick-open-workspace.ts:
```diff
@@ -75,12 +82,11 @@ export class WorkspaceQuickOpenItem extends QuickOpenItem {
      */
     getDescription(): string {
-        return (this.workspace) ? new URI(this.workspace).path.dir.toString() : '';
+        return (this.home) ? WorkspaceUtils.tildifyPath(this.workspace, this.home) : new URI(this.workspace).path.toString();
```
Just wondering, is it actually useful to have

    new URI(this.workspace).path.toString()

vs just

    this.workspace

?

vince-fugnitto replied:
the only reason I return `new URI(this.workspace).path.toString()` instead of `this.workspace` is that `this.workspace` returns the entire path including `file:///....` which I didn't want to necessarily display to end users.

akosyakov commented on packages/core/src/common/workspace-utils.ts:
```diff
@@ -0,0 +1,39 @@
+import { Path } from './path';
+import URI from './uri';
+
+export namespace WorkspaceUtils {
```
core should not have a mention of a workspace, there is a workspace package for it

vince-fugnitto replied:
I agree and the implementation does not require the context of a workspace anymore, perhaps I can move it into a new file called `path-utils` or I can directly add the function to the `path` file.

akosyakov replied:
`Path` is a set of generic functions to work with paths, this function seems to only make sense in the context of the filesystem. I would move it at least there.

akosyakov commented on packages/workspace/src/browser/quick-open-workspace.ts:
```diff
@@ -62,6 +68,7 @@ export class WorkspaceQuickOpenItem extends QuickOpenItem {
         private readonly workspaceService: WorkspaceService,
         private readonly messageService: MessageService,
         private readonly workspace: string,
+        private readonly home: string | undefined
```
How about precomputing `label` and `description` and passing them here. The super constructor already accepts them.

vince-fugnitto replied:
Just to clarify, you're suggesting to remove the subclass `WorkspaceQuickOpenItem` and directly use a precomputed `label`, `description` and `run` function when creating **items** for the `QuickOpenWorkspace`?
